File: 14-TypeScript/Grey Multi-Tenant/packages/grey-kotlin/src/main/kotlin/com/grey/sdk/domain/UserClient.kt
// =============================================================================
// Grey Multi-Tenant SDK - Kotlin
// domain/UserClient.kt
//
// Domain client for user operations.
// Wraps the gRPC service with a clean domain API.
// =============================================================================

package com.grey.sdk.domain

import com.grey.sdk.error.GreyError
import com.grey.sdk.error.GreyResult
import com.grey.sdk.grpc.UserGrpcService
import com.grey.sdk.grpc.UserResponse

/**
 * User data exposed to consumers.
 */
data class User(
    val id: String,
    val email: String,
    val name: String?,
    val avatarUrl: String?,
    val createdAt: Long,
    val updatedAt: Long
)

/**
 * Domain client for user operations.
 * All methods are suspend functions that return GreyResult.
 *
 * @property grpcService The underlying gRPC service
 * @property authClient The auth client for obtaining access tokens
 */
class UserClient internal constructor(
    private val grpcService: UserGrpcService,
    private val authClient: AuthClient
) {
    /**
     * Cached user data.
     */
    @Volatile
    private var _cachedUser: User? = null

    /**
     * The cached user, if available.
     */
    val cachedUser: User?
        get() = _cachedUser

    /**
     * Get the current user.
     * Requires authentication.
     *
     * @return GreyResult containing User or error
     */
    suspend fun getUser(): GreyResult<User> {
        val accessToken = authClient.accessToken
            ?: return GreyResult.failure(GreyError.unauthenticated())

        return grpcService.getUser(accessToken)
            .map { response ->
                val user = response.toUser()
                _cachedUser = user
                user
            }
    }

    /**
     * Get the current user, using cache if available.
     *
     * @param forceRefresh Force a fresh fetch from server
     * @return GreyResult containing User or error
     */
    suspend fun getUser(forceRefresh: Boolean): GreyResult<User> {
        if (!forceRefresh && _cachedUser != null) {
            return GreyResult.success(_cachedUser!!)
        }
        return getUser()
    }

    /**
     * Clear the cached user.
     */
    fun clearCache() {
        _cachedUser = null
    }

    /**
     * Extension to convert gRPC response to domain model.
     */
    private fun UserResponse.toUser(): User = User(
        id = id,
        email = email,
        name = name,
        avatarUrl = avatarUrl,
        createdAt = createdAt,
        updatedAt = updatedAt
    )
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-swift/Sources/GreySDK/Domain/UserClient.swift
// =============================================================================
// Grey Multi-Tenant SDK - Swift
// Domain/UserClient.swift
//
// Domain client for user operations.
// Wraps the gRPC service with a clean domain API.
// =============================================================================

import Foundation

// MARK: - Domain Models

/// User data exposed to consumers.
public struct User: Sendable {
    public let id: String
    public let email: String
    public let name: String?
    public let avatarUrl: String?
    public let createdAt: Int64
    public let updatedAt: Int64
    
    public init(
        id: String,
        email: String,
        name: String?,
        avatarUrl: String?,
        createdAt: Int64,
        updatedAt: Int64
    ) {
        self.id = id
        self.email = email
        self.name = name
        self.avatarUrl = avatarUrl
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}

// MARK: - Client

/// Domain client for user operations.
/// All methods are async functions that throw GreyError on failure.
public final class UserClient: @unchecked Sendable {
    
    private let grpcService: UserGRPCService
    private let authClient: AuthClient
    
    /// Cached user data.
    private var _cachedUser: User?
    
    /// Lock for thread-safe cache access.
    private let lock = NSLock()
    
    // MARK: - Initialization
    
    internal init(grpcService: UserGRPCService, authClient: AuthClient) {
        self.grpcService = grpcService
        self.authClient = authClient
    }
    
    // MARK: - Properties
    
    /// The cached user, if available.
    public var cachedUser: User? {
        lock.lock()
        defer { lock.unlock() }
        return _cachedUser
    }
    
    // MARK: - Operations
    
    /// Get the current user.
    /// Requires authentication.
    /// - Returns: User on success
    /// - Throws: GreyError on failure
    public func getUser() async throws -> User {
        guard let accessToken = authClient.accessToken else {
            throw GreyError.unauthenticated()
        }
        
        let response = try await grpcService.getUser(accessToken: accessToken)
        
        let user = User(
            id: response.id,
            email: response.email,
            name: response.name,
            avatarUrl: response.avatarUrl,
            createdAt: response.createdAt,
            updatedAt: response.updatedAt
        )
        
        lock.lock()
        _cachedUser = user
        lock.unlock()
        
        return user
    }
    
    /// Get the current user, using cache if available.
    /// - Parameter forceRefresh: Force a fresh fetch from server
    /// - Returns: User on success
    /// - Throws: GreyError on failure
    public func getUser(forceRefresh: Bool) async throws -> User {
        if !forceRefresh, let cached = cachedUser {
            return cached
        }
        return try await getUser()
    }
    
    /// Clear the cached user.
    public func clearCache() {
        lock.lock()
        defer { lock.unlock() }
        _cachedUser = nil
    }
}
